### DISCUSSION TITLE
Mediation coefficients

### DISCUSSION THREAD

**Discussion by swank10009** [raphaelvallat/pingouin#360] [Q&A]
Hello,
I have a question about the output from pingouin's mediation_analysis. Shouldn't the 'Indirect' coefficient be the product of the coefficients from 'M~X' and 'Y~M'? When I compare the output to that from the pyprocessmacro package, everything is the same except the stats for 'Y~M'. What is interesting is that 'Indirect' (the total indirect effects) is also the same, which suggests that the correct coefficient for 'Y~M' is being calculated. However, it doesn't seem like that is what is being returned. 
Please let me know if I am misinterpreting something.
Thanks

**Comment by swank10009** [comment]
I have no idea why the text above was struck through. That was not intentional.

**Comment by raphaelvallat** [comment]
Hi,

Can you please provide example scripts and data with `pingouin` and `pyprocessmacro`? The calculation of the indirect coefficient in Pingouin is done here:

https://github.com/raphaelvallat/pingouin/blob/5c5f61a09ee4e49ed81716fa9485527092fcc6b0/pingouin/regression.py#L952-L968

Please see this screenshot from Hayes and Rockwood 2017 for more details:

<img width="433" alt="image" src="https://github.com/raphaelvallat/pingouin/assets/25703792/237024c9-b902-4b16-978e-90ebc11dc0b7">

**Reply by swank10009** [reply]
_**Hello,
As is stated in the paragraph above from Hayes and Rockwood 2017, "...indirect effect is estimated as the product of regression coefficients a and b...".
Here is the code and output from pingouin:**_

_**Code:**_
`import pingouin as pg; df = pg.read_dataset('mediation'); pg.mediation_analysis(data=df, x='X', m='M', y='Y', seed=42, alpha=0.05, n_boot=1000)
`


_**Output:**_
       path      coef        se          pval  CI[2.5%]  CI[97.5%]  sig
0     M ~ X  0.561015  0.094480  4.391362e-08  0.373522   0.748509  Yes
1     Y ~ M  0.654173  0.085831  1.612674e-11  0.483844   0.824501  Yes
2     Total  0.396126  0.111160  5.671128e-04  0.175533   0.616719  Yes
3    Direct  0.039604  0.109648  7.187429e-01 -0.178018   0.257226   No
4  Indirect  0.356522  0.083984  0.000000e+00  0.218835   0.538517  Yes



**_In the output, the indirect path coefficients are 0.561015 and 0.654173, the product of which is 0.367. However, the total indirect coefficient is 0.35622._**


_**Using pyprocessmacro:**_

_**Code:**_
`from pyprocessmacro import Process; proc=Process(data=df, model=4, x='X', m=['M'], y='Y'); proc.summary()
`


_**Output:**_
***************************** OUTCOME MODELS ****************************

Outcome = Y 
OLS Regression Summary

     R²  Adj. R²    MSE       F  df1  df2  p-value
 0.3730   0.3534 2.6613 28.8524    2   97   0.0000

Coefficients

      coeff     se      t      p    LLCI   ULCI
Cons 1.9043 0.6055 3.1452 0.0022  0.7176 3.0909
X    0.0396 0.1096 0.3612 0.7187 -0.1753 0.2545
M    0.6355 0.1005 6.3212 0.0000  0.4385 0.8325

-------------------------------------------------------------------------

Outcome = M 
OLS Regression Summary

     R²  Adj. R²    MSE       F  df1  df2  p-value
 0.2646   0.2494 2.6868 35.2586    1   98   0.0000

Coefficients

           coeff     se      t      p   LLCI   ULCI
Cons 1.4995 0.5892 2.5450 0.0125 0.3447 2.6543
X    0.5610 0.0945 5.9379 0.0000 0.3758 0.7462

-------------------------------------------------------------------------


********************** DIRECT AND INDIRECT EFFECTS **********************

Direct effect of X on Y:

  Effect     SE      t      p    LLCI   ULCI
  0.0396 0.1096 0.3612 0.7187 -0.1753 0.2545

Indirect effect of X on Y:

  Effect  Boot SE  BootLLCI  BootULCI
  M  0.3565   0.0786    0.2191    0.5347


_**In the pyprocessmacro output, the indirect coefficients are 0.6355 and 0.5610, the product of which is 0.3565. This agrees to the total indirect effect shown at the bottom. It also agrees to the total indirect effect from pingouin (if you account for rounding). However, the coefficient for M in pingouin (0.654173) does not agree to that from pyprocessmacro (0.6355) and multiplying it by the other indirect coefficient does not agree to the total indirect effect. So, it seems like pingouin is calculating the correct coefficient for M (because the total indirect effect is correct), but the coefficient for M that is returned does not seem correct. I also tested output against cardiomoon's processR shiny app (https://cardiomoon.shinyapps.io/processR/). The app agrees to pyprocessmacro. Again, if I am just missing something, please let me know.**_

**Reply by raphaelvallat** [reply]
Hi,

Apologies about the delayed response. Thanks for the detailed post. So, if you look at the output of the pyprocessmacro the 0.6355 coefficient is the regression coefficient of Y ~ M **in a full model** that also includes X, i.e. Y ~ M + X

This is the same coefficient that Pingouin uses to calculate the indirect effect, here:

https://github.com/raphaelvallat/pingouin/blob/7923141161564b7a065b75f44f5fc75a2c1a1aa2/pingouin/regression.py#L952-L968

However, the `Y ~ M` coefficient reported in the Pingouin output table is **the univariate regression of M on Y, without the effect of X**, which I believe is not reported by pyprocessmacro.

In other words, Pingouin agrees with pyprocessmacro and uses the same coefficient internally, however, it reports the univariate regression coefficient and not the coefficient from the full model.

Let me know if that answered your question. I think it'd be a good idea to make this more explicit in the documentation or code perhaps.

Thanks
Raphael
